pkg/apis/core/validation/shoot.go

```diff
@@ -1938,14 +1946,84 @@ func validateShootOperationContext(operation string, shoot *core.Shoot, fldPath
+	return allErrs
+}
 
-	oldVal := oldMeta.Annotations[v1beta1constants.ShootAlphaControlPlaneHighAvailability]
+// ValidateShootHAControlPlaneUpdate validates the HA shoot control plane configuration
+func ValidateShootHAControlPlaneUpdate(newShoot, oldSnoot *core.Shoot) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// validate HA annotation if one exists and collect errors if any
+	allErrs = append(allErrs, validateShootHAControlPlaneAnnotationUpdate(newShoot.ObjectMeta, oldSnoot.ObjectMeta, field.NewPath("metadata"))...)
+	// validate HA ControlPlane Spec and collect errors if any
+	allErrs = append(allErrs, validateShootHAControlPlaneUpdate(&newShoot.Spec, &oldSnoot.Spec, field.NewPath("spec.controlPlane"))...)
+
+	return allErrs
+}
+
+func validateFailureToleranceValue(shoot *core.Shoot) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if annotationValue, ok := shoot.Annotations[v1beta1constants.ShootAlphaControlPlaneHighAvailability]; ok {
+		if !availableHAAnnotationValues.Has(annotationValue) {
+			allErrs = append(allErrs, field.NotSupported(field.NewPath("metadata", "annotations"), annotationValue, availableHAAnnotationValues.List()))
+		}
+	}
+	if shoot.Spec.ControlPlane != nil && shoot.Spec.ControlPlane.HighAvailability != nil {
+		failureToleranceType := string(shoot.Spec.ControlPlane.HighAvailability.FailureTolerance.Type)
+		if !availableFailureTolerance.Has(failureToleranceType) {
+			allErrs = append(allErrs, field.NotSupported(field.NewPath("spec", "controlPlane", "highAvailability", "failureTolerance", "type"), failureToleranceType, availableFailureTolerance.List()))
+		}
+	}
+	return allErrs
+}
+
+func bothHAAnnotationAndHASpecSet(shoot *core.Shoot) bool {
+	_, ok := shoot.Annotations[v1beta1constants.ShootAlphaControlPlaneHighAvailability]
+	return ok && shoot.Spec.ControlPlane != nil && shoot.Spec.ControlPlane.HighAvailability != nil
+}
+
+func validateShootHAControlPlaneAnnotationUpdate(newMeta, oldMeta metav1.ObjectMeta, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	oldVal, oldValExists := oldMeta.Annotations[v1beta1constants.ShootAlphaControlPlaneHighAvailability]
 	newVal := newMeta.Annotations[v1beta1constants.ShootAlphaControlPlaneHighAvailability]
 
-	// The etcd cluster cannot be scaled up/down nor is there an automatic re-scheduling to move from single- to multi-zone or the other way around.
-	allErrs = append(allErrs, apivalidation.ValidateImmutableField(oldVal, newVal, fldPath.Child("annotations").Key(v1beta1constants.ShootAlphaControlPlaneHighAvailability))...)
+	// If there is no annotation set then there is no need to compare if this has been changed.
+	// On the other hand if an old value exists for this annotation then enforce that it should
+	// not be changed.
+	if oldValExists {
+		allErrs = append(allErrs, apivalidation.ValidateImmutableField(newVal, oldVal, fldPath.Child("annotations").Key(v1beta1constants.ShootAlphaControlPlaneHighAvailability))...)
+	}
```

Comment:
This forbids that one can switch from the alpha annotation to spec.

Reply:
I never considered this case. Thanks for pointing it out. This means that i can no longer have 2 separate functions. It needs to be combined into one.

Reply:
pushed the changes

Reply:
This has been fixed